### DISCUSSION TITLE
TokioMdns takes a long time to discover peers

### DISCUSSION THREAD

**Discussion by Kelo007** [libp2p/rust-libp2p#2916] [Q&A] [answered]
I'm trying to build a tokio version mdns. The code is almost the same as `examples/mdns-passive-discovery.rs`.

It takes a long time to discover peers from start (about 10 secs).

<img width="1271" alt="pic" src="https://user-images.githubusercontent.com/12185797/191003309-23633320-15e8-4d24-93e5-95ac4a6f845a.png">

Is this normal or is there something wrong?

```rust
// Copyright 2018 Parity Technologies (UK) Ltd.
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

use futures::StreamExt;
use libp2p::{
    identity,
    mdns::{MdnsConfig, MdnsEvent, TokioMdns},
    swarm::{Swarm, SwarmEvent},
    PeerId,
};
use log::debug;
use std::error::Error;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    env_logger::init();

    // Create a random PeerId.
    let id_keys = identity::Keypair::generate_ed25519();
    let peer_id = PeerId::from(id_keys.public());
    println!("Local peer id: {:?}", peer_id);

    // Create a transport.
    let transport = libp2p::tokio_development_transport(id_keys)?;

    // Create an MDNS network behaviour.
    let behaviour = TokioMdns::new(MdnsConfig::default()).await?;

    // Create a Swarm that establishes connections through the given transport.
    // Note that the MDNS behaviour itself will not actually inititiate any connections,
    // as it only uses UDP.
    let mut swarm = Swarm::new(transport, behaviour, peer_id);
    swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse()?)?;

    loop {
        match swarm.select_next_some().await {
            SwarmEvent::Behaviour(MdnsEvent::Discovered(peers)) => {
                for (peer, addr) in peers {
                    debug!("discovered {} {}", peer, addr);
                }
            }
            SwarmEvent::Behaviour(MdnsEvent::Expired(expired)) => {
                for (peer, addr) in expired {
                    debug!("expired {} {}", peer, addr);
                }
            }
            _ => {}
        }
    }
}

```

**Comment by thomaseizinger** [comment]
Thanks for posting! I had a play around with this locally and can confirm that it _sometimes_ takes longer. After running the example repeatedly though, the discovery is basically instant.

I am not sure where that waiting is happening here to be honest.

**Comment by dariusc93** [comment]
Honestly thought it was me noticing this. Its rare though but at times do take about 5 to 10 seconds for it to discover but many times its instant. Could maybe test libp2p before commit https://github.com/libp2p/rust-libp2p/commit/89f898c69f6e58944e746565c3f8f8a3bbf3324e to see if it occurs there with tokio.

**Comment by Kelo007** [comment, marked as answer]
Actually in my local environment, it's not *sometimes* but *always*.

@dariusc93 I did a test. The discovery is instant before commit 89f898c.

**Reply by thomaseizinger** [reply]
Interesting, thanks for debugging!

I think this might be related to https://github.com/libp2p/rust-libp2p/pull/2748#issuecomment-1226527537. cc @gallegogt 

I think what might be happening is that the tokio implementation always waits for the first timer tick before discovering. Interestingly, the default query interval is 5 minutes so I am not sure where the 10 seconds delay comes from.

**Reply by gallegogt** [reply]
Hi everyone, I'm going to review to see how it can be improved.

**Reply by gallegogt** [reply]
Hi @thomaseizinger, @Kelo007  I found something:

The timeout check in the Iface poll (This lines https://github.com/libp2p/rust-libp2p/blob/master/protocols/mdns/src/behaviour/iface.rs#L242-L245), maybe can be moved to this line  (https://github.com/libp2p/rust-libp2p/blob/master/protocols/mdns/src/behaviour/iface.rs#L223). 

with this we make sure that the query is always sent instantly once the timeout has expired; otherwise you would have to wait for the poll to run again.

**Reply by thomaseizinger** [reply]
Thanks! That might explain things!

The implementation doesn't quite agree with [our coding guidelines](https://github.com/libp2p/rust-libp2p/blob/master/docs/coding-guidelines.md). I'll prepare a patch.

**Reply by thomaseizinger** [reply]
@Kelo007 Can you try your example with https://github.com/libp2p/rust-libp2p/pull/2939 included and check if that fixes it for you?

**Reply by Kelo007** [reply]
Yes, it is fixed! Thank you very much!
